controllers/kubeadm_control_plane_controller.go

```diff
@@ -104,24 +115,223 @@ func (r *KubeadmControlPlaneReconciler) Reconcile(req ctrl.Request) (res ctrl.Re
+	if err != nil {
+		logger.Error(err, "Failed to retrieve owner Cluster from the API Server")
+		return ctrl.Result{Requeue: true}
+	}
+	if cluster == nil {
+		logger.Info("Cluster Controller has not yet set OwnerRef")
+		return reconcile.Result{}
+	}
+	logger = logger.WithValues("cluster", cluster.Name)
+
+	labelSelector := generateKubeadmControlPlaneSelector(cluster.Name)
+	selector, err := metav1.LabelSelectorAsSelector(labelSelector)
+	if err != nil {
+		// Since we are building up the LabelSelector above, this should not fail
+		logger.Error(err, "failed to parse label selector")
+		return ctrl.Result{Requeue: true}
+	}
+	// Copy label selector to its status counterpart in string format.
+	// This is necessary for CRDs including scale subresources.
+	kcp.Status.Selector = selector.String()
+
+	selectorMap, err := metav1.LabelSelectorAsMap(labelSelector)
+	if err != nil {
+		// Since we are building up the LabelSelector above, this should not fail
+		logger.Error(err, "failed to convert label selector to a map")
+		return ctrl.Result{Requeue: true}
+	}
+
+	// Get all Machines linked to this KubeadmControlPlane.
+	allMachines := &clusterv1.MachineList{}
+	err = r.Client.List(
+		context.Background(),
+		allMachines,
+		client.InNamespace(kcp.Namespace),
+		client.MatchingLabels(selectorMap),
+	)
+	if err != nil {
+		logger.Error(err, "failed to list machines")
+		return ctrl.Result{Requeue: true}
+	}
+
+	// Filter out any machines that are not owned by this controller
+	// TODO: handle proper adoption of Machines
+	var ownedMachines []*clusterv1.Machine
+	for i := range allMachines.Items {
+		m := allMachines.Items[i]
+		controllerRef := metav1.GetControllerOf(&m)
+		if controllerRef != nil && controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == kcp.Name {
+			ownedMachines = append(ownedMachines, &m)
+		}
+	}
+
+	// Generate Cluster Certificates if needed
+	config := kcp.Spec.KubeadmConfigSpec.DeepCopy()
+	config.JoinConfiguration = nil
+	certificates := secret.NewCertificatesForInitialControlPlane(config.ClusterConfiguration)
+	if err := certificates.LookupOrGenerate(ctx, r.Client, cluster.Name, kcp.Namespace, *metav1.NewControllerRef(kcp, clusterv1.GroupVersion.WithKind("KubeadmControlPlane"))); err != nil {
+		logger.Error(err, "unable to lookup or create cluster certificates")
+		return ctrl.Result{Requeue: true}
+	}
+
+	// Generate Cluster Kubeconfig if needed
+	if kubeconfigErr := r.reconcileKubeconfig(ctx, cluster.Name, kcp.Namespace, cluster.Spec.ControlPlaneEndpoint, kcp); kubeconfigErr != nil {
+		if requeueErr, ok := errors.Cause(kubeconfigErr).(capierrors.HasRequeueAfterError); ok {
+			logger.Error(kubeconfigErr, "required certificates not found, requeueing")
+			return ctrl.Result{
+				Requeue:      true,
+				RequeueAfter: requeueErr.GetRequeueAfter(),
+			}
+		}
+		logger.Error(kubeconfigErr, "failed to reconcile Kubeconfig")
+		return ctrl.Result{Requeue: true}
+	}
+
```

Comment:
Should we short-circuit if cluster.Spec.ControlPlaneEndpoint is unset?

Reply:
Yes, updated.